pkg/controller/clone-controller.go

```diff
@@ -0,0 +1,260 @@
+		// A missing volume would most likely indicate a pod that has been created manually, but also incorrectly defined.
+		return errors.Errorf("Pod does not contain volume %q", DataVolName)
+	}
+
+	glog.V(Vdebug).Infof("processPodItem: getting pvc for key %q", pvcKey)
+	pvc, err := c.pvcFromKey(pvcKey)
+	if err != nil {
+		return errors.WithMessage(err, "could not retrieve pvc from cache")
+	}
+	// see if pvc's pod phase anno needs to be added/updated
+	phase := string(pod.Status.Phase)
+	if !checkIfAnnoExists(pvc, AnnPodPhase, phase) {
+		pvc, err = setPVCAnnotation(c.clientset, pvc, AnnCloneOf, "true")
+		if err != nil {
+			return errors.WithMessage(err, fmt.Sprintf("could not set annotation \"%s: %s\" on pvc %q", AnnPodPhase, phase, pvc.Name))
+			glog.V(Vdebug).Infof("processPodItem: pod phase %q annotated in pvc %q", pod.Status.Phase, pvcKey)
+		}
+	}
+	return nil
+}
+
+// Select only pvcs with the 'CloneRequest' annotation and that are not being processed.
+// We forget the key unless `processPvcItem` returns an error in which case the key can be
+// retried.
+func (c *CloneController) ProcessNextPvcItem() bool {
+	key, shutdown := c.pvcQueue.Get()
+	if shutdown {
+		return false
+	}
+	defer c.pvcQueue.Done(key)
+
+	pvc, err := c.pvcFromKey(key)
+	if err != nil || pvc == nil {
+		return c.forgetKey(key, fmt.Sprintf("ProcessNextPvcItem: error converting key %v to pvc: %v", key, err))
+	}
+	// filter pvc and decide if the source and target pods should be created
+	if continue_processing, _, _ := checkClonePVC(c.clientset, pvc, false); !continue_processing {
+		return c.forgetKey(key, fmt.Sprintf("ProcessNextPvcItem: skipping pvc %q\n", key))
+	}
+	glog.V(Vdebug).Infof("ProcessNextPvcItem: next pvc to process: %s\n", key)
+	err = c.processPvcItem(pvc)
+	if err != nil { // processPvcItem errors may not have been logged so log here
+		glog.Errorf("error processing pvc %q: %v", key, err)
+		return true
+	}
+	return c.forgetKey(key, fmt.Sprintf("ProcessNextPvcItem: processing pvc %q completed", key))
+}
+
+// Create the cloning source and target pods based the pvc. The pvc is checked (again) to ensure that we are not already
+// processing this pvc, which would result in multiple pods for the same pvc.
+func (c *CloneController) processPvcItem(pvc *v1.PersistentVolumeClaim) error {
+	cr, err := getCloneRequestPVC(pvc)
+	if err != nil {
+		return err
+	}
+	
+	// check our existing pvc one more time to ensure we should be working on it
+	// and to help mitigate any race conditions. This time we get the latest pvc.
+	doCreate, pvc, err := checkClonePVC(c.clientset, pvc, true)
+	if err != nil { // maybe an intermittent api error
+		return err
+	}
+	if !doCreate { // don't create pods but not an error
+		return nil // forget key; logging already done
+	}
+
+	// update pvc with 'AnnCloningPods' annotation to indicate cloning is in process
+	anno := map[string]string{AnnCloningPods: "exist"}
```

Comment:
should "exists" be a constant?

Reply:
no value is really needed here, just the anno key. Same would apply to createdByController and its value of "yes".